chore: 1031 migrate docs to nuxt UI docs template (#1049)
Tresjs/tres: docs/content/3.api/1.components/tres-objects.md

---
title: Tres Components
description: Learn how TresJS automatically maps Vue components to Three.js objects using a custom renderer and autogenerated catalogue.
---

> The core idea of **Tres** is an _autogenerated catalogue_ of all the Three.js elements. This catalogue is generated from the Three.js source code, so it's always up to date.

Reviewer: ```suggestion
The core idea of **TresJS** is to provide an _autogenerated catalogue_ of all the Three.js elements. This catalogue is generated based on the Three.js source code, so it's always up to date.
```

When using plain Three.js, you need to import the elements you want to use. For example, if you want to use a `PerspectiveCamera`, you need to import it from the `three` package:

```js
import { PerspectiveCamera } from 'three'

const camera = new PerspectiveCamera(45, width / height, 1, 1000)
```
::tip{icon="i-ph-lightbulb-duotone"}

> With **Tres** you don't need to import anything, that's because **Tres** automatically generate a **Vue Component based of the Three Object you want to use in PascalCase with a Tres prefix**.

Reviewer: ```suggestion
With **Tres** you don't need to import anything, that's because **Tres** automatically generates a **Vue Component based of the Three Object you want to use in PascalCase with a Tres prefix**.
```

::

> For example, if you want to use a `PerspectiveCamera` you would use `<TresPerspectiveCamera />` component.

Reviewer: ```suggestion
For example, if you want to use a `PerspectiveCamera` you would use the `<TresPerspectiveCamera />` component.
```

```vue
<template>
  <TresCanvas>
    <TresPerspectiveCamera />
    <!-- Your scene goes here -->
  </TresCanvas>
</template>
```

This means that you can rely on the same [documentation](https://threejs.org/docs/?q=Perspe#api/en/cameras/PerspectiveCamera) as you would when using plain Three.js, but with the power of Vue.

## Declaring objects

If we follow this argument, you should be able to lay out an instance like this: ❌

```vue
<template>
  <TresCanvas>
    <TresPerspectiveCamera visible :position="new THREE.Vector3(1, 2, 3)" />
    <!-- Your scene goes here -->
  </TresCanvas>
</template>
```
But with Tres this is not needed, you can define properties declaratively like this: ✅

```vue
<template>
  <TresCanvas>
    <TresPerspectiveCamera visible :position="[1, 2, 3]" />
    <!-- Your scene goes here -->
  </TresCanvas>
</template>
```

> TresJS follows a simple naming convention: any component prefixed with `Tres` is automatically mapped to its Three.js counterpart:

Reviewer: **suggestion(non-blocking):**
```suggestion
TresJS follows a simple naming convention: any component prefixed with `Tres` is automatically mapped to its Three.js counterpart. Some examples:
```

| Vue Component | Instance |
|---------------|-------------|
| `<TresPerspectiveCamera />` | `new THREE.PerspectiveCamera()` |
| `<TresMesh />` | `new THREE.Mesh()` |
| `<TresBoxGeometry />` | `new THREE.BoxGeometry()` |
| `<TresMeshBasicMaterial />` | `new THREE.MeshBasicMaterial()` |

## Constructor Arguments with `args`

Many Three.js objects require constructor arguments. TresJS provides the `args` prop as an **array of constructor arguments** that are passed directly to the Three.js constructor:

For example, the `PerspectiveCamera` constructor has the following arguments:

> - `fov` - Camera frustum vertical field of view.
> - `aspect` - Camera frustum aspect ratio.
> - `near` - Camera frustum near plane.
> - `far` - Camera frustum far plane.

Reviewer: ```suggestion
- `fov` - Camera frustum vertical field of view
- `aspect` - Camera frustum aspect ratio
- `near` - Camera frustum near plane
- `far` - Camera frustum far plane
```
